### PR TITLE
Implementing ``any`` intrinsic

### PR DESCRIPTION
**Pull request by czgdp1807**
Here's how plan to do it,

Since `axis` is a runtime argument to the `any` intrinsic, the only thing we know for certain is the range which it can take, `[1, n_dims - 1]` where `n_dims` is the rank of input `mask`. So, in an `if-else-if` ladder, consider all the possible values in `[1, n_dims - 1]` and generate nested do loops and allocate result array for each case. So for any `axis` correct do loops and result array will be allocated. Let me know if you know of a better way. I brainstormed and figured this out.

### CODE REVIEW

File: src/libasr/pass/intrinsic_function_registry.h [hunk 1]
```diff
@@ -646,6 +646,24 @@ namespace IntrinsicFunctionRegistry {
...
     }
 
+    static inline bool is_elemental(int64_t id) {
+        ASRUtils::IntrinsicFunctions id_ = static_cast<ASRUtils::IntrinsicFunctions>(id);
+        return ( id_ == ASRUtils::IntrinsicFunctions::Abs ||
+                 id_ == ASRUtils::IntrinsicFunctions::Cos ||
+                 id_ == ASRUtils::IntrinsicFunctions::Gamma ||
+                 id_ == ASRUtils::IntrinsicFunctions::LogGamma ||
+                 id_ == ASRUtils::IntrinsicFunctions::Sin );
+    }
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:656]: This is fine for now. Should we have a new member "elemental" in `IntrinsicFunction`?

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -646,6 +646,24 @@ namespace IntrinsicFunctionRegistry {
...
+
+    static inline int get_dim_index(ASRUtils::IntrinsicFunctions id) {
+        if( id == ASRUtils::IntrinsicFunctions::Any ) {
+            return 1;
+        } else {
+            LCOMPILERS_ASSERT(false);
+        }
+        return -1;
+    }
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:665]: What does this function do? Write some documentation as a comment.

**Comment by certik** [comment]
Looks good so far. Once this is done, it will setup the infrastructure for all the array functions (sum, maxval, etc.).

File: src/libasr/pass/intrinsic_function_registry.h (hunk 1 above)
Comment (czgdp1807) [reply src/libasr/pass/intrinsic_function_registry.h:656]: Definitely we should have.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -461,6 +535,281 @@ namespace Abs {
...
+
+    return ASR::make_IntrinsicFunction_t(al, loc,
+        static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Any),
+        args.p, args.n, overload_id, logical_return_type, value);
+}
+
+static inline void generate_body(Allocator& al, const Location& loc,
+    ASR::expr_t* array, ASR::expr_t* return_var, SymbolTable* fn_scope,
+    Vec<ASR::stmt_t*>& fn_body) {
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:602]: I would distinguish these by name, not function overloading. I think it's clearer that way.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 2 above)
Comment (czgdp1807) [reply src/libasr/pass/intrinsic_function_registry.h:885]: Done.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 3 above)
Comment (czgdp1807) [reply src/libasr/pass/intrinsic_function_registry.h:602]: Done.

**Comment by czgdp1807** [comment]
@certik Let me know if this looks good to you. I will clean up the git history and merge in that case.

File: integration_tests/modules_36.f90 [hunk 4]
```diff
@@ -0,0 +1,31 @@
+module fpm_main_01
```
Comment (certik) [inline integration_tests/modules_36.f90:1]: You need to register this test I think.

**Review by certik — APPROVED** [review]
I think this looks great! Thanks.

**Comment by czgdp1807** [comment]
Okay. After registering the test and cleaning up the git history I will merge the PR and start working on `sum`.

**Comment by certik** [comment]
@czgdp1807 perfect! Thank you, this will unblock a lot of efforts. After any and sum is done, we can easily do all the others I think.

File: integration_tests/modules_36.f90 (hunk 4 above)
Comment (czgdp1807) [reply integration_tests/modules_36.f90]: Done.

File: src/libasr/asr_verify.cpp [hunk 5]
```diff
@@ -857,15 +857,20 @@ class VerifyVisitor : public BaseWalkVisitor<VerifyVisitor>
...
                     ASR::is_a<ASR::FunctionType_t>(*ASRUtils::symbol_type(fn))) ||
                     ASR::is_a<ASR::ClassProcedure_t>(*fn),
                 "FunctionCall::m_name must be a Function or Variable with FunctionType");
         }
 
+        if( fn && ASR::is_a<ASR::Function_t>(*fn) ) {
+            ASR::Function_t* fn_ = ASR::down_cast<ASR::Function_t>(fn);
+            require(fn_->m_return_var != nullptr,
+                    "FunctionCall::m_name must be returning a non-void value.");
+        }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/asr_verify.cpp:873]: @czgdp1807, this fails for the variables with struct as a return type.
As we don't have void type, should I create a variable with integer type, assign zero to it and use this as a return variable?

Comment (Thirumalai-Shaktivel) [reply src/libasr/asr_verify.cpp:873]: Example program that doesn't work:
```fortran
program expr2
implicit none

type t
    integer :: x
end type t

type(t) :: a
a = test()
print *, a%x

contains
    function test() result(p)
        type(t) :: p
        p%x = 1
    end function test
end program
```

Comment (certik) [reply src/libasr/asr_verify.cpp:873]: @Thirumalai-Shaktivel please open up a new issue for this. Let's fix it.

Comment (Thirumalai-Shaktivel) [reply src/libasr/asr_verify.cpp:873]: Created an issue here: https://github.com/lfortran/lfortran/issues/1593

**Pull request merged**